### PR TITLE
feat: monitor certificate expiration

### PR DESCRIPTION
**Pull request by RomainMuller**
Raise high severity alarms when certificates are within 45 days of their
expiration (at which point ACM should already have renewed them). This
can be opted out using domain configuration.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: src/monitoring/monitored-certificate.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+   /**
+    * The name of the CloudWatch metric to be emitted.
+    *
+    * @default 'DaysToExpiry'
+    */
+   readonly metricName?: string;
+}
+
+export class MonitoredCertificate extends Construct implements IMonitoredCertificate {
```
Comment (eladb) [inline src/monitoring/monitored-certificate.ts:38]: docstring?

File: src/monitoring/monitored-certificate.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  private readonly customMetricName: string;
+  private readonly customMetricNamespace: string;
+
+  public constructor(scope: Construct, id: string, private readonly props: MonitoredCertificateProps) {
+    super(scope, id);
+
+    this.customMetricName = props.metricName ?? 'DaysToExpiry';
```
Comment (eladb) [inline src/monitoring/monitored-certificate.ts:45]: why "custom"?

File: src/monitoring/monitored-certificate.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    this.customMetricNamespace = props.metricNamespace ?? Stack.of(this).stackName;
+
+    const dynamicMonitor = new CertificateMonitor(this, `${props.certificate.node.addr}::Monitor`, {
```
Comment (eladb) [inline src/monitoring/monitored-certificate.ts:48]: Redundant. Scope name is intrinsically part of the ID:

```suggestion
    const dynamicMonitor = new CertificateMonitor(this, `Monitor`, {
```

File: src/monitoring/monitored-certificate.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      .createAlarm(this, 'EndpointAlarm', {
+        alarmDescription: `The certificate used to serve ${props.domainName} will expire in less than 45 days!`,
+        comparisonOperator: ComparisonOperator.LESS_THAN_THRESHOLD,
+        evaluationPeriods: 1,
+        threshold: 45,
+        treatMissingData: TreatMissingData.BREACHING,
+      });
+    endpointAlarm.node.addDependency(dynamicMonitor);
+
+    props.monitoring.addHighSeverityAlarm('ACM Certificate will expire soon!', acmAlarm);
```
Comment (eladb) [inline src/monitoring/monitored-certificate.ts:78]: I would move the coupling to ConstructHub's `Monitoring` class _out_ of this construct (think of this construct as something we'd want to publish as a separate library) and just expose the alarms as properties and bind them to the alarms in the `ConstructHub`

File: src/monitoring/index.ts [hunk 5]
```diff
@@ -74,6 +77,23 @@ export class Monitoring extends Construct {
...
+   * This evaluates both the metric of the ACM certificate, and the expiration
+   * date of the certificate used by the provided HTTPS endpoint.
+   *
+   * @param certificate the certificate to monitor
+   * @param domainName  the endpoint to monitor
+   *
+   * @returns the `MonitoredCertificate` instance.
+   */
+  public addMonitoredCertificate(certificate: acm.ICertificate, domainName: string, opts: MonitoredCertificateOptions = {}): IMonitoredCertificate {
+    return new MonitoredCertificate(this, `${certificate.node.addr}::${domainName}`, { ...opts, certificate, domainName, monitoring: this });
```
Comment (eladb) [inline src/monitoring/index.ts:94]: The idea was that `Monitoring` is pretty general-purpose. I would move this code `WebApp` (alongside the code that binds the alarms to high severity actions).

**Review by eladb — CHANGES_REQUESTED** [review]

File: src/monitoring/monitored-certificate.ts (hunk 2 above)
Comment (RomainMuller) [reply src/monitoring/monitored-certificate.ts:45]: This metric is emitted by the Lambda function, so is a "custom" metric in that sense... It's not the AWS-provided metric for the ACM certificate.

File: src/monitoring/monitored-certificate.ts (hunk 3 above)
Comment (RomainMuller) [reply src/monitoring/monitored-certificate.ts:48]: Oh yeah - traces of a refactor :P

File: src/monitoring/monitored-certificate.ts (hunk 2 above)
Comment (eladb) [reply src/monitoring/monitored-certificate.ts:45]: The fact that it is "custom" is an implementation leak.

From an API perspective, it is a metric emitted by this construct. The fact that it is "custom" is an implementation detail.

**Review by eladb — APPROVED** [review]

**Pull request merged**